### PR TITLE
Add execution statistic warnings

### PR DESCRIPTION
**Pull request by knwg145**
Statistic warnings are generated based on execution stats and is a way of "looking back to see what might have gone wrong".

A new warning collector has been created to collect warnings from different phases of the query. It uses a map of : Map<Warningcode, Set> 

in order to store multiple warnings of the same code and to get rid of duplicates. It can also be applied to different phases (since it's a set).

Open to discussion about this.

I've included 3 execution-stastics based warnings in this PR

- Skew Warning 
Warns about task skews in each stage

- CPU and Memory Threshold warning
Trigger a warning if the peak Memory or peak CPU uses more than a certain threshold (configurable) value for the maximum memory or cpu.

- Better Join Ordering
Triggers a warning if there is a better order for a join

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/execution/warnings/statswarnings/JoinWarningsGenerator.java [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+                if (operatorStat.getPlanNodeId().equals(planNode.getId())) {
+                    return operatorStat.getOutputPositions();
+                }
+            }
+            return 0;
+        }
+
+        private void generateJoinWarnings(WarningSource left, WarningSource right, PlanNode leftNode, PlanNode rightNode, String joinLabel, String stageId)
+        {
+            long leftNumberOfRow = getOperatorstats(leftNode);
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/execution/warnings/statswarnings/JoinWarningsGenerator.java]: `getOperatorstats` -> `getOperatorStats`

File: presto-main/src/main/java/io/prestosql/execution/warnings/statswarnings/SkewWarningsGenerator.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+import io.prestosql.execution.TaskInfo;
+import io.prestosql.spi.PrestoWarning;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static io.prestosql.spi.connector.StandardWarningCode.STAGE_SKEW;
+
+public class SkewWarningsGenerator
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/execution/warnings/statswarnings/SkewWarningsGenerator.java]: rename to `CPUSkewWarningsGenerator` as tasks might also be skewed on positions

File: presto-main/src/main/java/io/prestosql/execution/warnings/statswarnings/JoinWarningsGenerator.java [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+                rightMessageBuilder.add(format("Tablename: %s", right.getTableNames().toString()));
+            }
+            if (!right.getStageIds().isEmpty()) {
+                rightMessageBuilder.add(format("Stage IDs: %s", right.getStageIds().toString()));
+            }
+
+            leftMessage = String.join(",", leftMessageBuilder);
+            rightMessage = String.join(",", rightMessageBuilder);
+
+            String warning = format("Try swapping the join ordering for: (%s) %s (%s) in StageId: %s",
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/execution/warnings/statswarnings/JoinWarningsGenerator.java]: You could also check if stats are available for join inputs and suggest using CBO instead.

File: presto-main/src/main/java/io/prestosql/execution/warnings/statswarnings/JoinWarningsGenerator.java [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static io.prestosql.spi.connector.StandardWarningCode.BETTER_JOIN_ORDERING;
+import static java.lang.String.format;
+
+public class JoinWarningsGenerator
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/execution/warnings/statswarnings/JoinWarningsGenerator.java]: You could also generate warning (to use CBO) when build side is tiny, join input stats are available and replicated join is used.

**Comment by colebow** [comment]
:wave: @knwg145 - this PR is inactive and doesn't seem to be under development. If you'd like to continue work on this at any point in the future, feel free to re-open.

**Pull request closed without merging**